### ISSUE TITLE
sync dying on special characters

### ISSUE THREAD

**Issue by scottw-finao** [aws/aws-cli#5050]
I'm assuming the characters are mac characters or perhaps foreign letters as we have a number of mac using customers and some international customers.

When I do a sync it finishes with a number of messages like:

```
M_DC_ENGASAMP/KM_DC_SAMP_004.jpg Unable to parse response (reference to invalid character number: line 3, column 3), invalid XML received:
b'<?xml version="1.0" encoding="UTF-8"?>\n<InitiateMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Bucket>finao-images</Bucket><Key>16109/KM_DC_SAMPENGALB\rIKM&#x8;&#x8;&#x8;\rKM_DC_ENGASAMP/KM_DC_SAMP_004.jpg</Key><UploadId>N6hAkCuL84Qaz0eXoofdm434tIkJedfLwyxJMuaHg0rGvN0h689G7Y5c9u3iDjsITpRdtrDN5gsqdy7skNgG15dhO6HkQsI4WNBJBZGGl4axR_F_WLFrBpNlPx1lKyTx</UploadId></InitiateMultipartUploadResult>'
KM_DC_ENGASAMP/KM_DC_SAMP_009.jpg Unable to parse response (reference to invalid character number: line 3, column 3), invalid XML received:
b'<?xml version="1.0" encoding="UTF-8"?>\n<InitiateMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Bucket>finao-images</Bucket><Key>16109/KM_DC_SAMPENGALB\rIKM&#x8;&#x8;&#x8;\rKM_DC_ENGASAMP/KM_DC_SAMP_009.jpg</Key><UploadId>3MQHZas84HaksvIZRcogcDbXxtYdSxi.L2y4Ra9ptCCxZJcus3K2KuCYeSJ_Yx7eljzPozZAloAUsXH2F.k1YHa575RWT6VWRbelHYZrVXvJ95GrRm_0G8pZ2KQtCIpC</UploadId></InitiateMultipartUploadResult>'
KM_DC_ENGASAMP/KM_DC_SAMP_003.jpg Unable to parse response (reference to invalid character number: line 3, column 3), invalid XML received:
b'<?xml version="1.0" encoding="UTF-8"?>\n<InitiateMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Bucket>finao-images</Bucket><Key>16109/KM_DC_SAMPENGALB\rIKM&#x8;&#x8;&#x8;\rKM_DC_ENGASAMP/KM_DC_SAMP_003.jpg</Key><UploadId>nyu5d_10_lc1niXppHGYQk_yjWHY8lamA64321APp22NcTTRGXzeUqyvp9SITRisgPfjUxqPQIFMnOX8dLjLLsVhl8pnMgRr_Q8mCVEds7EIP.QLTT456Db.k3zxGaok</UploadId></InitiateMultipartUploadResult>'
upload failed: ../../srv/ftp/17325/Preview of \udce2__Untitled Book\udce2__.jpg to s3://finao-images/17325/Preview of \udce2__Untitled Book\udce2__.jpg 'utf-8' codec can't encode character '\udce2' in position 17: surrogates not allowed
upload failed: ../../srv/ftp/39369/Promenade \udcc3_ Cheval.jpg to s3://finao-images/39369/Promenade \udcc3_ Cheval.jpg 'utf-8' codec can't encode character '\udcc3' in position 16: surrogates not allowed
upload failed: ../../srv/ftp/50709/8x10-The children photography with natural light in San Antonio\udcef__TX-9-web-seo-k.jpg to s3://finao-images/50709/8x10-The children photography with natural light in San Antonio\udcef__TX-9-web-seo-k.jpg 'utf-8' codec can't encode character '\udcef' in position 69: surrogates not allowed
upload failed: ../../srv/ftp/50709/8x10-The children photography with natural light in San Antonio\udcef__TX-web-seo-key.jpg to s3://finao-images/50709/8x10-The children photography with natural light in San Antonio\udcef__TX-web-seo-key.jpg 'utf-8' codec can't encode character '\udcef' in position 69: surrogates not allowed
upload failed: ../../srv/ftp/50709/Black and white fine art children photography with natural light in San Antonio\udcef.jpg to s3://finao-images/50709/Black and white fine art children photography with natural light in San Antonio\udcef.jpg 'utf-8' codec can't encode character '\udcef' in position 85: surrogates not allowed
upload failed: ../../srv/ftp/50709/The children photography with natural light in San Antonio\udcef__TX-4-web-seo-keywor.jpg to s3://finao-images/50709/The children photography with natural light in San Antonio\udcef__TX-4-web-seo-keywor.jpg 'utf-8' codec can't encode character '\udcef' in position 64: surrogates not allowed
upload failed: ../../srv/ftp/50709/The children photography with natural light in San Antonio\udcef__TX-9-web-seo-keywor.jpg to s3://finao-images/50709/The children photography with natural light in San Antonio\udcef__TX-9-web-seo-keywor.jpg 'utf-8' codec can't encode character '\udcef' in position 64: surrogates not allowed
upload failed: ../../srv/ftp/50709/The children photography with natural light in San Antonio\udcef__TX-web-seo-keywords.jpg to s3://finao-images/50709/The children photography with natural light in San Antonio\udcef__TX-web-seo-keywords.jpg 'utf-8' codec can't encode character '\udcef' in position 64: surrogates not allowed
```

local for the system uploading is ubuntu set to en_US.UTF-8

```
$ locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

```

**Comment by scottw-finao** [comment]
the first few errors look like it was a improperly created folder (there were ^H characters in the name so it looked like someone was trying to backspace and change the name, but it sent the backspace characters to the filesystem somehow)
I corrected that one. But the others all appear to be user generated filenames of one sort or another.

**Comment by klaytaybai** [comment]
It looks like your file keys were not originally encoded in UTF-8.  If you can figure out why that is, it might be the best place to solve your problem.  The [S3 key naming guidelines](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html) give more guidance around this topic.  They say to use UTF-8 and avoid characters such as '\' and '^'.

**Comment by scottw-finao** [comment]
All told, there only ended up being about <20 files and most of them appeared to be from 2014 or before. (we have periodic programs that trim out old files, and these predate the cutoff by more than about 5-6 years, so they apparently break the cleaning code also)
As a result, the files I found are well beyond our 'still needed' dating so I just renamed them, replacing special characters with underscores (how our more recent code handles anything that doesn't match the known-allowed characters on our various systems) so I don't think we'll be encountering them again any time soon.

I just figured I would point them out as things like special mac characters and even back-space characters do manage to sidestep locale entries from time to time, and part of the reason they created problems for me is that even the 'snowball' import we did seemed to copy the files as they were.  (I ended up having to delete the entire path that had the corrupted folder name in it as I couldn't find a prefix syntax that would allow me to just delete objects starting with the beginning of the folder name - I then had to re-sync everything under the pseudo-parent folder to restore the other files that were not corrupt.
